lib/src/conflicts.rs

```diff
@@ -469,30 +482,47 @@ pub async fn update_from_content(
 
     // If the user edited the empty placeholder for an absent side, we consider the
     // conflict resolved.
-    if zip(contents.iter(), file_ids.iter())
+    if zip(contents.iter(), used_file_ids.iter())
         .any(|(content, file_id)| file_id.is_none() && !content.is_empty())
     {
         let file_id = store.write_file(path, &mut &content[..])?;
         return Ok(Merge::normal(file_id));
     }
 
     // Now write the new files contents we found by parsing the file with conflict
-    // markers. Update the Merge object with the new FileIds.
-    let builder: BackendResult<MergeBuilder<Option<FileId>>> =
-        zip(contents.iter(), file_ids.iter())
-            .map(|(content, file_id)| {
-                match file_id {
-                    Some(_) => {
-                        let file_id = store.write_file(path, &mut content.as_slice())?;
-                        Ok(Some(file_id))
-                    }
-                    None => {
-                        // The missing side of a conflict is still represented by
-                        // the empty string we materialized it as
-                        Ok(None)
-                    }
+    // markers.
+    let mut new_file_ids: Vec<Option<FileId>> = zip(contents.iter(), used_file_ids.iter())
+        .map(|(content, file_id)| -> BackendResult<Option<FileId>> {
+            match file_id {
+                Some(_) => {
+                    let file_id = store.write_file(path, &mut content.as_slice())?;
+                    Ok(Some(file_id))
                 }
+                None => {
+                    // The missing side of a conflict is still represented by
+                    // the empty string we materialized it as
+                    Ok(None)
+                }
+            }
+        })
+        .try_collect()?;
+
+    // If the conflict was simplified, expand the conflict to the original
+    // number of sides.
+    if new_file_ids.len() != file_ids.num_sides() * 2 - 1 {
```

Comment:
You could have something like:

```rust
    if new_file_ids.len() != file_ids.iter().len() {
        assert_eq!(new_file_ids.len(), simplified_file_ids.iter().len());
        ...
    }
```

Moreover, as you are generating `new_file_ids`, you could have generated a mapping from `used_file_ids` to `new_file_ids` (probably only the non-None values) and use it here. I'm not sure that's always more efficient, or that this is the best solution possible, but it would never be very slow, and I'm a bit uneasy about unnecessarily quadratic algorithms.